syz-manager/rpc.go

```diff
@@ -262,161 +176,150 @@ func (serv *RPCServer) Check(a *rpctype.CheckArgs, r *int) error {
 	serv.mgr.machineChecked(a, serv.targetEnabledSyscalls)
 	a.DisabledCalls = nil
 	serv.checkResult = a
-	serv.rotator = prog.MakeRotator(serv.cfg.Target, serv.targetEnabledSyscalls, serv.rnd)
 	return nil
 }
 
-func (serv *RPCServer) NewInput(a *rpctype.NewInputArgs, r *int) error {
-	p, disabled, bad := parseProgram(serv.cfg.Target, serv.targetEnabledSyscalls, a.Input.Prog)
-	if bad != nil || disabled {
-		log.Errorf("rejecting program from fuzzer (bad=%v, disabled=%v):\n%s", bad, disabled, a.Input.Prog)
+func (serv *RPCServer) ExchangeInfo(a *rpctype.ExchangeInfoRequest, r *rpctype.ExchangeInfoReply) error {
+	fuzzer := serv.mgr.getFuzzer()
+	if fuzzer == nil {
+		// The fuzzer is not initialized yet.
 		return nil
 	}
-	serv.mu.Lock()
-	defer serv.mu.Unlock()
+	serv.stats.mergeNamed(a.StatsDelta)
 
-	f := serv.fuzzers[a.Name]
-	if f != nil {
-		a.Cover, a.Signal = f.instModules.Canonicalize(a.Cover, a.Signal)
+	serv.mu.Lock()
+	runner := serv.runners[a.Name]
+	serv.mu.Unlock()
+	if runner == nil {
+		// There might be a parallel shutdownInstance().
+		// Ignore the request then.
+		return nil
 	}
-	inputSignal := a.Signal.Deserialize()
 
-	inp := corpus.NewInput{
-		Prog:   p,
-		Call:   a.Call,
-		Signal: inputSignal,
-		Cover:  a.Cover,
+	for _, result := range a.Results {
+		runner.doneRequest(result, fuzzer)
 	}
 
-	log.Logf(4, "new input from %v for syscall %v (signal=%v, cover=%v)",
-		a.Name, inp.StringCall(), inputSignal.Len(), len(a.Cover))
-	// Note: f may be nil if we called shutdownInstance,
-	// but this request is already in-flight.
-	genuine := !serv.corpusSignal.Diff(inputSignal).Empty()
-	rotated := false
-	if !genuine && f != nil && f.rotated {
-		rotated = !f.rotatedSignal.Diff(inputSignal).Empty()
-	}
-	if !genuine && !rotated {
-		return nil
-	}
-	if !serv.mgr.newInput(inp) {
-		return nil
+	for i := 0; i < a.NeedProgs; i++ {
+		inp := fuzzer.NextInput()
+		id := serv.nextRequestID.Add(1)
+		r.Requests = append(r.Requests, runner.newRequest(id, inp))
 	}
 
-	if f != nil && f.rotated {
-		f.rotatedSignal.Merge(inputSignal)
-	}
-	diff := serv.corpusCover.MergeDiff(a.Cover)
-	serv.stats.corpusCover.set(len(serv.corpusCover))
-	if len(diff) != 0 && serv.coverFilter != nil {
-		// Note: ReportGenerator is already initialized if coverFilter is enabled.
-		rg, err := getReportGenerator(serv.cfg, serv.modules)
-		if err != nil {
-			return err
-		}
-		filtered := 0
-		for _, pc := range diff {
-			if serv.coverFilter[uint32(rg.RestorePC(pc))] != 0 {
-				filtered++
-			}
-		}
-		serv.stats.corpusCoverFiltered.add(filtered)
-	}
-	serv.stats.newInputs.inc()
-	if rotated {
-		serv.stats.rotatedInputs.inc()
-	}
+	runner.mu.Lock()
```

Comment:
If we don't do parallel requests from the VM, there should be no contention. Otherwise I'd definitely try to put `fuzzer.NextInput()` requests out of the critical section.